### PR TITLE
Only allow (op/type)defs to be added by resource

### PR DESCRIPTION
**Pull request by ss2165**

### CODE REVIEW

File: src/resource.rs [hunk 1]
```diff
@@ -581,39 +569,85 @@ impl Resource {
...
+            name,
+            description,
+            params,
+            misc,
+            SignatureFunc::CustomFunc(Box::new(signature_func)),
+        )
+    }
+
+    /// Create an OpDef with a signature (inputs+outputs) read from the
+    /// declarative TAML
```
Comment (ss2165) [inline src/resource.rs:634]:
```suggestion
    /// declarative YAML
```

Comment (cqc-alec) [inline src/resource.rs:634]:
```suggestion
    /// declarative YAML
```

File: src/resource.rs [hunk 2]
```diff
@@ -565,14 +539,28 @@ impl Resource {
...
 
     /// Allows read-only access to the types in this Resource
-    pub fn types(&self) -> &HashMap<SmolStr, TypeDef> {
-        &self.types
+    pub fn get_type(&self, type_name: &str) -> Option<&TypeDef> {
+        self.types.get(type_name)
+    }
+
+    /// Allows read-only access to the operations in this Resource
+    pub fn get_op_mut(&mut self, op_name: &str) -> Option<&mut Arc<OpDef>> {
```
Comment (cqc-alec) [inline src/resource.rs:562]: Is this method useful?

File: src/extensions/logic.rs [hunk 3]
```diff
@@ -29,65 +29,67 @@ pub fn resource() -> Resource {
...
-            };
-            Ok((
-                vec![bool_type(); n as usize].into(),
-                vec![bool_type()].into(),
-                ResourceSet::default(),
-            ))
-        },
-    );
+    resource
+        .add_op_custom_sig(
```
Comment (cqc-alec) [inline src/extensions/logic.rs:33]: Could we just use `add_op()` for this?

File: src/extensions/rotation.rs [hunk 4]
```diff
@@ -26,21 +26,22 @@ pub const fn resource_id() -> SmolStr {
...
-            Ok((t.clone(), t, ResourceSet::default()))
-        },
-    );
-
-    resource.add_op(op).unwrap();
+    Type::Angle.add_to_resource(&mut resource);
+    Type::Quaternion.add_to_resource(&mut resource);
+
+    resource
+        .add_op_custom_sig(
```
Comment (cqc-alec) [inline src/extensions/rotation.rs:33]: Could we just use `add_op()` for this?

File: src/resource.rs [hunk 5]
```diff
@@ -565,14 +539,28 @@ impl Resource {
...
 
+    /// Returns the number of operations of this [`Resource`].
+    pub fn num_operations(&self) -> usize {
+        self.operations.len()
+    }
+
+    /// Returns the number of types of this [`Resource`].
+    pub fn num_types(&self) -> usize {
+        self.types.len()
+    }
```
[the remark this answers is not in the document]
Comment (ss2165) [reply src/resource.rs:550]: I only added the ones used by tests

File: src/resource.rs (hunk 2 above)
Comment (ss2165) [reply src/resource.rs:562]: Used by `OpDef::add_lowering`

File: src/extensions/logic.rs (hunk 3 above)
Comment (ss2165) [reply src/extensions/logic.rs:33]: the `SignatureFunc` enum is private (which makes sense to me, given the serialisation trickery used internally)

File: src/resource.rs (hunk 2 above)
Comment (cqc-alec) [reply src/resource.rs:562]: Hmm, can't find `get_op_mut` used anywhere on this branch at the moment?

Comment (ss2165) [reply src/resource.rs:562]: sorry I mean, in order for `add_lowering` to be used, you would need a mutable reference. I guess I can add a test for that function.

Comment (cqc-alec) [reply src/resource.rs:562]: Ah, right I see!

**Review by cqc-alec — APPROVED** [review]

File: src/resource.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (ss2165) [reply src/resource.rs]: gone now

File: src/resource.rs [hunk 6]
```diff
@@ -565,14 +539,28 @@ impl Resource {
...
     /// Allows read-only access to the operations in this Resource
-    pub fn operations(&self) -> &HashMap<SmolStr, Arc<OpDef>> {
```
[the remark this answers is not in the document]
Comment (ss2165) [reply src/resource.rs:569]: added

**Comment by ss2165** [comment]
@cqc-alec  re-requesting as I've made some major edits

took to opportunity to refactor TypeDef and OpDef out, you can glance over those commits

**Pull request merged**